src/io/peek.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Read wrapper that allows peeking ahead in the stream without consuming
+// the peeked bytes.  BufRead.fill_buf() does not provide this, since it
+// only guarantees to return one byte.  For simplicity, we implement this
+// as a thin wrapper around BufReader.
+
+use bytes::{Buf, BytesMut};
+use std::io::{BufRead, BufReader, Read, Result, Seek, SeekFrom};
+
+pub struct PeekReader<R: Read> {
```

Comment:
I did read through this Reader code and it seems sane, but I may have missed some edge cases. The testcases seems pretty extensive, so I mostly trust them for this logic.